### ISSUE TITLE
Add a general-purpose CSGConvexHull3D node

### ISSUE THREAD

**Issue by 31** [godotengine/godot-proposals#9933]
### Describe the project you are working on

A game that includes a 3d modeling tool using CSG, where "fill all the space between this set of points" is a common operation, and the result must be displayed as a mesh.

### Describe the problem or limitation you are having in your project

It's not trivial to create a triangle mesh from a 3d convex hull. It's hard to find a library that properly handles common edge cases (faces with more than 3 vertices, which occurs in something as simple as a common cube). Godot has some features available that use convex hulls (especially collision), but none do meshing.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

`CSGConvexHull3D`, a simple CSG node that has `points` and makes it into a convex hull mesh compatible with the rest of the CSG system.

* There's no need for a nice editor GUI for setting up `points` in my case because I will be using this node programmatically.
* For people who will use it in the editor, they may want to write their own Tool script that adds functionality on top, to make it easy to use for specific common shapes in their particular game. (However, an integrated `points` editor might be useful for very organic and one-off shapes.)

> [!NOTE]
> Although I personally don't need a nice editor GUI, it is considered a requirement by the Godot team. https://github.com/godotengine/godot/pull/100725#issuecomment-2567104880, https://github.com/godotengine/godot/pull/100725#issuecomment-2562998485

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

PR https://github.com/godotengine/godot/pull/94321 integrates Manifold, which seems to have good convex hull meshing support. I wrote up a draft `CSGConvexHull3D` node implementation based on that PR here: https://github.com/godotengine/godot/commit/25cc9b0fca89fe2100a76ffb461ac18f8325d3dd.

I'm not familiar with https://github.com/godotengine/godot/blob/5833f597865c773fae3ee09fc4e31d4a243f812d/core/math/convex_hull.h and if it creates good meshes, but it's not usable from game code. If the API were exposed, it might be reasonable to use this to generate a mesh to pass into `CSGMesh3D`. I suspect that would be relatively slow.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

In theory I could create a 3d convex hull mesh by bringing in some other library or implementation and then pass the result into `CSGMesh3D`, but there are reasons I don't want to try:
* Godot already has convex hull logic. Any additional library would be extra weight.
* It wouldn't be able to benefit from any optimizations that can happen due to keeping the hull in a more "native" format.
* It's (generally speaking) difficult to get this kind of code working in all exports as an asset/GDExtension rather than as a module/fork.

---

Related:

* https://github.com/godotengine/godot-proposals/issues/9711
  * https://github.com/godotengine/godot/pull/91748
  * https://github.com/godotengine/godot/pull/94321
* https://github.com/godotengine/godot/pull/100725

**Comment by Calinou** [comment]
Does this *have* to be a CSG mesh? I'm concerned about adding more CSG meshes without an equivalent PrimitiveMesh, as this can encourage users to use slow CSG functionality when it's not needed. We have the same problem with CSGPolygon3D's Path mode, which lacks a PrimitiveMesh equivalent.

**Comment by 31** [comment]
> Does this _have_ to be a CSG mesh?

No, but the cost is (potential) performance. That's what this is about from the proposal:

> > [Passing a convex hull mesh into the CSGMesh3D node] wouldn't be able to benefit from any optimizations that can happen due to keeping the hull in a more "native" format.

Internally, the CSG system could avoid creating a Godot mesh until it has applied all the CSG operations using Manifold, and avoid some pointless conversions back and forth. (*Does* it do this? Maybe not, but it seems like fairly obvious potential to me. 😄)

> I'm concerned about adding more CSG meshes without an equivalent PrimitiveMesh, as this can encourage users to use slow CSG functionality when it's not needed. We have the same problem with CSGPolygon3D's Path mode, which lacks a PrimitiveMesh equivalent.

That's reasonable, I have nothing against also adding meshes for this, and for path-mode polygons as well.

Another factor: it was super easy to add the CSG node because all the functionality is right there for interop with Manifold. I'm not sure it would be so easy with meshes. For practical reasons I want to file a proposal with a ready implementation to try to get it done sooner rather than waiting for someone who knows how to implement it as a mesh, which could never happen.

**Comment by fire** [comment]
I support this proposal. Hulling and keeping the promise of manifold, is not trivial to do. I think manifold is the primary FOSS library that can do this.

**Comment by fire** [comment]
It still needs to have a points editor by default.

Would you have any thoughts on using @celyk 's points editor plugin? I'll poke for more details.

**Comment by celyk** [comment]
Hello! I recently made a plugin called [VertexHandles](https://github.com/celyk/godot-useful-stuff/tree/main/addons/VertexHandles) that allows you to edit the vertices of a Mesh when placed under a MeshInstance3D. The code could be modified to generate a mesh based on the points instead. I hope this helps!

https://github.com/user-attachments/assets/c509b3be-b005-4c79-ba3b-94d36a6e3033

**Comment by fire** [comment]
Ready for implementation!

**Comment by 31** [comment]
I brought the old commit back up to speed and submitted a PR: https://github.com/godotengine/godot/pull/100725. The shape mesh itself doesn't detect clicks in the editor, which I think is fine for my own purposes, but maybe it's an indication of a deeper problem, so I've left it as a draft for now.

If someone wants to take this notification as impetus to work on a Mesh version of a convex hull rather than CSG node, or otherwise do something better than my PR, I'd be grateful. 😄

**Comment by 31** [comment]
> Hello! I recently made a plugin called [VertexHandles](https://github.com/celyk/godot-useful-stuff/tree/main/addons/VertexHandles) that allows you to edit the vertices of a Mesh when placed under a MeshInstance3D. The code could be modified to generate a mesh based on the points instead. I hope this helps!

Thanks @celyk! I made [some updates](https://gist.github.com/31/ac9ebdfc9b702ec135b3446d63d7a86e) to make it work with the PR node and posted a video at https://github.com/godotengine/godot/pull/100725.

**Comment by 31** [comment]
Now that 4.4 has come out with Manifold-based CSG (awesome to see, congrats!) I figured I would update this proposal.

@fire and @Calinou clarified in my PR https://github.com/godotengine/godot/pull/100725 that a point editor would be required for this feature to make it into Godot. I can't do it (it dramatically increases the scope of the work, in my opinion) so I closed the PR in Jan 2025.

I'm closing this proposal now, too--my intent was to make a proposal that was useful, easy to implement, and could be improved upon later, but that didn't work out. I do still think this would be a valuable feature, but to be accepted, it seems there are a few things that would need to be addressed in a new proposal by someone who knows Godot a lot better than I do:

* Include a design for a point editor in the proposal.
* Make it a type of mesh that can optionally be put into the CSG system rather than a type of CSG node.

---

Working on this did teach me a lot about how Manifold's APIs work, even though I didn't end up with the Godot feature I was aiming for. I hope I was more helpful than nuisance along the way with the overall Manifold work. 😄
